core/src/main/java/org/apache/iceberg/avro/SchemaToType.java

```diff
@@ -104,13 +104,23 @@ public Type record(Schema record, List<String> names, List<Type> fieldTypes) {
 
   @Override
   public Type union(Schema union, List<Type> options) {
-    Preconditions.checkArgument(
-        AvroSchemaUtil.isOptionSchema(union), "Unsupported type: non-option union: %s", union);
-    // records, arrays, and maps will check nullability later
-    if (options.get(0) == null) {
-      return options.get(1);
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      if (options.get(0) == null) {
+        return options.get(1);
+      } else {
+        return options.get(0);
+      }
     } else {
-      return options.get(0);
+      // Create list of Iceberg schema fields
+      List<Types.NestedField> fields = Lists.newArrayListWithExpectedSize(options.size());
+      int tagIndex = 0;
+      fields.add(Types.NestedField.required(allocateId(), "tag", Types.IntegerType.get()));
+      for (Type option : options) {
+        if (option != null) {
```

Comment:
I'm thinking through the case of a non-option union with null, `["int", "null", "float"]`. That would produce: `struct<1: tag required int, 2: field0 optional int, 3: field1 optional float>`.

When the union's value is null, the would be `row(tag=1, field0=null, field1=null)` and when the union's value is a float it would be `row(tag=2, field0=null, field1=3.1)`. The mismatch between field names and the `tag` value is odd to me. I think the tag index should be updated every time through the loop, regardless of whether the option type is `null`.

Also, since these are called options, what about using "option" rather than "field"?

Reply:
To clarify:

1- Does this depend on whether there is a `null` option or not? I think the behavior is the same. The current logic creates `field_i` corresponding to non-null union options, starting from 0. The behavior sounds the same whether `null` is there (in the options) or not. This means that `tag` convention does not depend on whether there is a `null` option or not. Do you agree?

2- `tag` is set at read time. In the read PR, it is set in such a way that it is zero-based index as well. The test cases are [here](https://github.com/apache/iceberg/pull/5704/files#diff-409a0a443647e843af30276987f07ddd067c2ef7660d632e89bfce0d783c18e3R94). So in both cases, whether there is a null option or not, it will match `i` in `field_i`.

3- I think when the union value is null, the struct will be null as well (although this does not affect the two points above).

Reply:
My understanding was that the intent here was to skip any NULL entry in the union. That would mean that the example above has tags that don't match the `fieldN` naming convention after any null value.

However, it looks like this is checking for Java's `null` rather than `Schema.create(Schema.Type.NULL)`.

If you're not skipping the NULL schema in ID assignment, then I'm happy with it.

Reply:
So this is skipping the optional branch in the union. In this step, its value is Java's null (see the counterpart simple union if branch  -a few lines before this- which also uses Java's null).

Reply:
I don't think we should skip the optional branch. That makes the field names no longer aligned with the tag.

Reply:
The union null branch is to denote that the complex union as optional. The corresponding field will always be null, and even when the union is null, the whole corresponding struct is null (so it will not be the case that all fields in the struct are null). Typically that is also how other engines go about union to struct conversion. Here are some examples in [Hive](ttps://github.com/apache/hive/blob/4f1dba31b25fb2c3b071492ff06b54f76a55c99a/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFExtractUnion.java#L193) and [Trino](https://github.com/trinodb/trino/blob/1621cb6d92b3e96b8204ba71ceca06e8090f0a58/plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveTypeTranslator.java#L217) (they convert from Hive schema but it is always nullable). Also Spark `SchemaConverters` [explicitly drops the null before mapping to struct](https://github.com/apache/spark/blob/a5c53384def22b01b8ef28bee6f2d10648bce1a1/connector/avro/src/main/scala/org/apache/spark/sql/avro/SchemaConverters.scala#L129).

We would want to be compatible with other other conventions because:
* It actually makes sense to view the union type as optional + the main types, so the indexes map to the main types.
* Continue the convention adopted elsewhere and not create multiple conventions/standards.
* Avoid adding data conversion adapters (between two schema conventions) during migrations from other formats to Iceberg, especially in the Trino case, where the schemas exactly match (`(tag, field_0, field_1,...)` with no dedicated null field). Trino set the precedent that converts union to struct at the table level (unlike Hive which uses explicit UDF, and Spark that does it at the file level). In addition to avoiding migrations, we could establish the schema field names/convention adopted by both Iceberg and Trino as the standard for other engines to follow.

Reply:
Also, the `tag` field value will align with the non-optional branch indexes (in subsequent read PR).

Reply:
I think I'm okay with this to match the behavior of other conversions. We do need to ensure that the tag matches the field name. That's going to be annoying in the read path.